### PR TITLE
Issue #1962: Fix flexbox grid system negatively affects float fallbac…

### PR DESCRIPTION
**Pull request by jenlampton**
PR for https://github.com/backdrop/backdrop-issues/issues/1962

Replaces https://github.com/backdrop/backdrop/pull/1442

### CODE REVIEW

**Comment by herbdool** [comment]
Tested with a browser testing tool with IE11. The third column is still getting pushed below. Firebug shows that the fallback css isn't getting added.

Used Backdrop 1.4.2 with core Bartik/Geary

**Comment by jenlampton** [comment]
Interesting. I tested in IE10 and it worked yesterday. Maybe IE 11 doesn't
trigger the fallback? @wesruv any insight?

File: core/modules/layout/css/grid-float.css [hunk 1]
```diff
@@ -141,6 +146,11 @@
   padding-right: 1rem;
   padding-left: 1rem;
   float: left;
+  /* Override flexbox values in case they're set */
+  -webkit-box-flex: unset;
+  -webkit-flex: unset;
+  -ms-flex: unset;
```
Comment (quicksketch) [inline core/modules/layout/css/grid-float.css]: This also needs to be `none` for `-ms-flex`.

File: core/modules/layout/css/grid-float.css [hunk 2]
```diff
@@ -950,3 +960,63 @@
...
+.col-sm-last,
+.col-md-first,
+.col-md-last,
+.col-lg-first,
+.col-lg-last,
+.col-xl-first,
+.col-xl-last {
+  -webkit-box-ordinal-group: unset;
+  -webkit-order: unset;
+  -ms-flex-order:unset;
```
Comment (quicksketch) [inline core/modules/layout/css/grid-float.css]: `-ms-flex-order` apparently doesn't support `unset`, but you could put it back to the default value of `0`.

File: core/modules/layout/css/grid-float.css [hunk 3]
```diff
@@ -950,3 +960,63 @@
...
+.row-md-bottom,
+.row-lg-top,
+.row-lg-center,
+.row-lg-bottom,
+.row-xl-top,
+.row-xl-center,
+.row-xl-bottom {
+  -webkit-box-align: unset;
+  -webkit-align-items: unset;
+  -ms-flex-align: unset;
```
Comment (quicksketch) [inline core/modules/layout/css/grid-float.css]: The MS documentation doesn't say it supports either `unset` or `none` for this property, though I'm not sure it matters. The default value is `stretch` according to https://msdn.microsoft.com/en-us/library/hh673531%28v=vs.85%29.aspx

File: core/modules/layout/css/grid-float.css [hunk 4]
```diff
@@ -950,3 +960,63 @@
...
+.col-md-center,
+.col-md-bottom,
+.col-lg-top,
+.col-lg-center,
+.col-lg-bottom,
+.col-xl-top,
+.col-xl-center,
+.col-xl-bottom {
+  -webkit-align-self: unset;
+  -ms-flex-item-align: unset;
```
Comment (quicksketch) [inline core/modules/layout/css/grid-float.css]: The default `-ms-flex-item-align` is `auto`, which we'd need to use as IE10 doesn't support `unset`.

File: core/modules/layout/css/grid-float.css (hunk 1 above)
Comment (jenlampton) [inline core/modules/layout/css/grid-float.css]: fixed.

File: core/modules/layout/css/grid-float.css (hunk 2 above)
Comment (jenlampton) [inline core/modules/layout/css/grid-float.css]: fixed.

File: core/modules/layout/css/grid-float.css (hunk 3 above)
Comment (jenlampton) [inline core/modules/layout/css/grid-float.css]: fixed.

File: core/modules/layout/css/grid-float.css (hunk 4 above)
Comment (jenlampton) [inline core/modules/layout/css/grid-float.css]: fixed.

**Comment by jenlampton** [comment]
Updated PR based on feedback, and rebased.

**Comment by jenlampton** [comment]
Well, I'm testing again in browserstack and I'm still having some issues with IE11. The problem is that we have `display: -ms-flexbox;` defined before `display: flex;` for `.row` and the `display: flex;` breaks in IE11.  Remove that (or set it back to `-ms-flexbox;`) and everything's fine.

**Comment by quicksketch** [comment]
The latest PR looks great. I tested it in IE10 and IE11 and both seem to work now.

> Tested with a browser testing tool with IE11. The third column is still getting pushed below. Firebug shows that the fallback css isn't getting added.

IE10+ supports flexbox, so it won't add the fallback CSS and it should just use the flexbox styling that is there. The problem seemed to have come from IE10/11 having a bug in the calculation of widths when dealing flex elements: http://stackoverflow.com/questions/21942183/multiline-flexbox-in-ie11-calculating-widths-incorrectly

Apparently IE12 this bug has been fixed, but in the mean time, the solution that works for grid-like situations like this is simple to add a `max-width` property. The arrangement of the page can then be handled by flexbox, but the width that is calculated incorrectly by IE10/11 is still caught by `max-width`, making columns fit within the row instead of pushing down.

**Comment by quicksketch** [comment]
I've merged this in for 1.4.3, thanks @jenlampton!

**Comment by herbdool** [comment]
Looks good in IE now, thanks! In Safari it's still pushing the last column.

**Comment by jenlampton** [comment]
I m having some problems I. Safari with Moscow layout (sidebar disappears)
we will likely need to push another CSS fix shortly.
- -
  Please excuse brevity and typos. Message typed with my thumbs from an Open
  Source OS.

**Comment by Graham-72** [comment]
I have reported my problems with Safari at https://github.com/backdrop/backdrop-issues/issues/1985

**Pull request merged**